### ISSUE TITLE
Error: Packets out of order. Got: 1 Expected: 2

### ISSUE THREAD

**Issue by adrianseeley** [mysqljs/mysql#1363]
I'm trying to do multi-row inserts using:

``` js
sql.query('INSERT INTO fx (t, s, o, h, l, c) VALUES ?', [values], function (err) {
```

I get:

```
Error: Packets out of order. Got: 1 Expected: 2
    at Parser.write (C:\Users\Admin\Desktop\fx\node_modules\mysql\lib\protocol\Parser.js:39:19)
    at Protocol.write (C:\Users\Admin\Desktop\fx\node_modules\mysql\lib\protocol\Protocol.js:39:16)
    at Socket.<anonymous> (C:\Users\Admin\Desktop\fx\node_modules\mysql\lib\Connection.js:96:28)
    at Socket.emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at TCP.onread (net.js:538:20)
    --------------------
    at Pool.query (C:\Users\Admin\Desktop\fx\node_modules\mysql\lib\Pool.js:191:23)
    at non_recurse_l (C:\Users\Admin\Desktop\fx\fx_to_db.js:50:6)
    at recurse_d (C:\Users\Admin\Desktop\fx\fx_to_db.js:67:9)
    at Object.<anonymous> (C:\Users\Admin\Desktop\fx\fx_to_db.js:70:8)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
```

As my stack trace, I did some playing around, and inserting 1000 or so rows works perfectly fine in a single query, but trying to insert 100k or even 1m starts throwing packets out of order errors.

Package version from NPM: `2.10.2`
MySql version: `MySQL Workbench Community (GPL) for Windows version 6.3.6 CE build 511 (64 bit)`

**Comment by dougwilson** [comment]
Hi @adrianseeley, what version of Node.js are you using?

**Comment by adrianseeley** [comment]
The most up to date at the time of writing:
```
Node.js v5.7.1		2016-03-02	4.6.85.31	3.6.0	47
```

**Comment by dougwilson** [comment]
Thank, @adrianseeley! I've been trying to track this down, though several issues, and I haven't yet been able to pinpoint the reason. No one so far has shared code that reproduces, though. Would you be able to help out in that regard? Provide some DDL to make the table and then a simple file that does enough to cause that error to happen?

**Comment by adrianseeley** [comment]
That code is long gone now, but I can try to replicate it - just give me some time.

**Comment by dougwilson** [comment]
Cool, no problem. I'll continue to take a look, so don't bend over backwards to get it if it's not easy :)

**Comment by adrianseeley** [comment]
My pleasure to help out @dougwilson, can't report a problem if you aren't willing to help fix it!

```javascript
/*

CREATE TABLE `bugtest`.`fx` (
  `id` INT NOT NULL AUTO_INCREMENT,
  `t` BIGINT NOT NULL,
  `s` VARCHAR(6) NOT NULL,
  `o` FLOAT NOT NULL,
  `h` FLOAT NOT NULL,
  `l` FLOAT NOT NULL,
  `c` FLOAT NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE INDEX `id_UNIQUE` (`id` ASC));

SET GLOBAL max_allowed_packet=1073741824;
Select @@global.max_allowed_packet; 

  */


var mysql = require('mysql');
var sql   = mysql.createPool({
    connectionLimit: 10,
	host:            'localhost',
	user:            'root',
	password:        'redacted',
	database:        'bugtest'
});

var values = [];

for (var i = 0; i < 1000000; i ++) {
	values.push([new Date().getTime(), 'EURUSD', Math.random(), Math.random(), Math.random(), Math.random()]);
}

sql.query('INSERT INTO fx (t, s, o, h, l, c) VALUES ?', [values], function (err, res) {
	if (err) {
		throw err;
	}
	console.log(res);
});
```

Built this little script with some SQL setup notation at the top, I even tried expanding the max packet size to allow for a million rows inserted at once and nothing seemed to crash it. This is using a more recent version of node `v5.10.1`. Looks like the node update may have fixed something in there?

**Comment by dougwilson** [comment]
Thanks so much! Yea, I'm not sure. I tried on 5.0.0, 5.10.1, 5.7.1 and a couple more and either got OOM errors or it worked fine, so not there error here. With the various reports every so often, it almost seems like it could be an issue with this module, though I have never personally encountered it. I haven't gone far enough to try against some specific MySQL versions that were reported to see if that's playing a role yet.

**Comment by adrianseeley** [comment]
No worries, let me know if there is anything else I can do to help the process along. There should be a way to increase the node heap size, but that's probably out of scope for real world usage.

**Comment by dougwilson** [comment]
Thanks, @adrianseeley! If you can try the new 2.11.0 to see if you're still encountering the issue, that would help a lot as well!

**Comment by alager** [comment]
I just updated to 2.11.1 and am getting this error when the DB daemon is shutdown.
The program flow is as follows for **each** request:
1. call `createConnection()` if the connection object doesn't exist
2. call `connection.connect()`
3. call `connection.query()`
4. in the query callback, call `connection.end()`

Now, after step 4 is complete, if I shut down the DB, then I get the exception.
My understanding is that the `end()` method should have closed the connection, so that it doesn't matter if the DB is there or not.

Is my program flow correct?
Is my assumption about what `end()` does correct?
I tried putting an error handler on it:  `connection.on( 'error'...` but it doesn't seem to be catching it.  Where would I put the error handler to catch this?

Thanks!

**Comment by dougwilson** [comment]
Hi @alager, the `.end()` method is actually sending a `END` packet sequence to the server, and the server actually has to take action and respond back, so no, it doesn't seem like your understanding is correct. If you want to simply sever the TCP connection, use `.destroy()`.

Using `connection.on( 'error'...` should work, so long as you did that before making the calls, not afterwards (since this is how events in Node.js work). If you provide the code I can use to reproduce the issue, I could probably say more.

**Comment by alager** [comment]
My mistake, that error handler does work, I just had not transferred the most recent version of that file, that contained the error handler code.

Using `.destroy()` I get the same result.
See my actual code below.  It's a little more complicated than the simple workflow, since I'm iterating over 15 different DBs to monitor them.  Which is why I need to gracefully handle DBs coming and going.

Thanks for your help and input!

PS. while I was putting this together, the DBs timed out the connections as 'idle'
`db error { [Error: Connection lost: The server closed the connection.] fatal: true, code: 'PROTOCOL_CONNECTION_LOST' }`
So they are definitely "hanging around"
```
 myMysql module
 // load up the connection array
 router.init = function()
 {
	arrayOfClusters.forEach( function( serverObj )
	{
		for( var server in serverObj )
		{
			var con  = mysql.createConnection(serverObj[server]);
			con.type = serverObj[server].type;
			con.clstName = serverObj[server].clstName;
			router.connection.push(con);
		}
	});

};

// connect to those pesky databases
router.connect = function( callback )
{
	var len = router.connection.length;
	var myParallelTasks = [];

	for( var idx = 0; idx < len; idx++ )
	{
		(function( index )
		{
			_dbConnector( index, callback );
		})(idx);
	}

	// execute the queries in parallel, then land here
	async.parallel( myParallelTasks, function()
	{
		// all done?
		callback();
	});

};


function _dbConnector( idx )
{
	(function( index )
	{
		router.connection[ index ].connect( function( err )
		{
			if( !err )
			{
				console.log("Database is connected ... \n\n");
			} else
			{
				console.log("*** Error connecting database *** \n" + err);
			}
		});

		// // setup an event handler to reconnect to the DB on disconnect
		router.connection[ index ].on( 'error', function( err )
		{
			console.log( 'db error', err );
		});
	})( idx );
}
// load up the connection array
router.init = function()
{
	arrayOfClusters.forEach( function( serverObj )
	{
		for( var server in serverObj )
		{
			var con  = mysql.createConnection(serverObj[server]);
			con.type = serverObj[server].type;
			con.clstName = serverObj[server].clstName;
			router.connection.push(con);
		}
	});
};

// connect to those pesky databases
router.connect = function( callback )
{
	var len = router.connection.length;
	var myParallelTasks = [];

	for( var idx = 0; idx < len; idx++ )
	{
		(function( index )
		{
			_dbConnector( index, callback );
		})(idx);
	}

	// execute the queries in parallel, then land here
	async.parallel( myParallelTasks, function()
	{
		// all done?
		callback();
	});

};


 function _dbConnector( idx )
 {
	(function( index )
	{
		router.connection[ index ].connect( function( err )
		{
			if( !err )
			{
				console.log("Database is connected ... \n\n");
			} else
			{
				console.log("*** Error connecting database *** \n" + err);
				//router.connection[ index ].destroy();
				
				// try again in 20 seconds
				//setTimeout( _dbConnector( index ), 20000);
				//process.exit(1);
			}
		});

		// // setup an event handler to reconnect to the DB on disconnect
		router.connection[ index ].on( 'error', function( err )
		{
			console.log( 'db error', err );
		});
	})( idx );
 }
```

```
request route:

router.get('/getInfo', function(req, res, next)
 {
 	if( dbConn.hasOwnProperty( "connection" ) )
 	{
 		if( dbConn.connection.length === 0 )
 		{
 			dbConn.init();
 		}
 		else
 		{
 			delete dbConn.connection;
 			dbConn.connection = [];
 			dbConn.init();
 		}
 	}
 	else
 		res.json({}); //return null if not set up correctly
 	
	dbConn.connect( function()
	{
		_pollEachDB( req, res, next );
	});
});
	

var _pollEachDB = function( req, res, next )
{
	var dbData = new dbObject();
	var sql = null;
	var clstName = null;
	
	if( req.query.hasOwnProperty( "clstName" ) )
		clstName = req.query.clstName;
	else
		res.json({}); //return null if no query passed in
	
	// get the DB up count
	if( dbConn.connection[0].type === "galeraCluster" )
	{
		sql = "show status like 'wsrep_%'";
	}

	var idx = 0;
	var len = dbConn.connection.length;
	var myParallelTasks = [];
	for( idx; idx <  len; idx ++ )
	{
		if( dbConn.connection[ idx ].clstName === clstName )
		{
			// wrap a closure around the callback so that
			// we can have our array index in the callback
			(function(index)
			{
				myParallelTasks.push(function(callback)
				{
					dbConn.connection[index].query(sql, function(err, data, fields)
					{
						dbConn.connection[index].destroy();
						if( !err )
						{
							var xformData             = _xformResults(data);
							//return the rows as JSON
							dbData.dbInfoArray.push( xformData );
							callback();
						}
						else
						{
							// dbConn.connection[index].end();
							console.log("SQL Error: " + err);
							dbData.dbInfoArray.push( null );
							callback();
						}
						
					});
				});
			})(idx);
		}
	}

	// execute the queries in parallel, then land here 
	async.parallel( myParallelTasks, function()
	{
		// all done?
		res.json( dbData.dbInfoArray );
	});
}
```

**Comment by frogcjn** [comment]
How to solve this problem?

**Comment by frogcjn** [comment]
I got `SequelizeDatabaseError: Packets out of order. Got: 1 Expected: 2`
if insert 10000 records

**Comment by dougwilson** [comment]
Hi @frogcjn, what version of this module are you using?

**Comment by frogcjn** [comment]
2.11.1

**Comment by dougwilson** [comment]
Thanks, @frogcjn! Do you think you can help move this bug forward? For example, would it be possible to provide all the necessary information such that I can resolve this issue or would you be willing to investigate and work on a PR to fix the issue?

**Comment by frogcjn** [comment]
I think the problem is trying to request as many as possible at the same time. Both request and http from node will report errors.

**Comment by frogcjn** [comment]
I use a retryIfError written by my self and limit the request count in concurrent may solve the problem for most cases.

**Comment by frogcjn** [comment]
I have 10000 records to insert at the same time. About 8 - 10 columns?

**Comment by frogcjn** [comment]
Now, I received `max_allowed_packet`

**Comment by kgdiem** [comment]
I guess I have some work to do if I'm going to insert 1134319 records :octocat:

**Comment by kgdiem** [comment]
@dougwilson is there any information in particular that you'd like in order to solve the issue? I'll be working with this dataset for a bit so I can reproduce the issue and do whatevs.

**Comment by sidorares** [comment]
@kgdiem http://dev.mysql.com/doc/refman/5.7/en/load-data.html might be a better option for you, much less overhead and does not have to fit in single mysql packet

**Comment by dougwilson** [comment]
Awesome to hear, @kgdiem ! It basically boils down to either (a) a pull request to fix the issue or (b) enough, detailed information for me to set everything up on my side in order to reproduce the issue (typically version of Node.js, version of this module, version of MySQL server, a complete JavaScript file I can copy and paste and run, and instructions on how to run the file).

**Comment by kgdiem** [comment]
Update on this, I wasn't able to reproduce the issue I was getting before. 

After changing my script around a bit I'm able to insert ~ 975,000 rows in a single query. The limiting factor there was the amount of memory node takes up with a string that long.  Probably fine there.

Otherwise, I assume you're already aware, but when rapidly running queries it's totally unpredictable as to how many records will be inserted. ie, how the package responds to classic DDoS behavior. 

For example, this code sometimes inserts 200, sometimes close to 600 before running out of memory (after ~ 500,000 calls):

```
                                 //DB connection  //Read stream
function insertPOSData(connection, stream){
  //start constructing the query
  const insertStr = 'INSERT INTO posTrans (basketNumber, pharmacyNumber, productNumber, salesDate, extSalesAmount, salesQty) VALUES ';

  let insert,
  i = 0;

  stream
  .pipe(parse())
  .on('data', (data) => {
    //add to the string 
    insert = insertStr + `("${data[0]}", "${data[1]}", "${data[2]}", "${data[3]}", "${data[4]}", "${data[5]}");`;
    i++;
    console.log(i);
    connection.query(insert);
  })
  .on('end', 
    () => {
      console.log("Success: POS Transactions");
      process.exit();
    })
  );
}
```

Probably something the user should be handling outside of this mysql module anyways. Just reporting my findings. 

Sorry I couldn't have been of more assistance. If you'd like the code I was using let me know.

**Comment by dougwilson** [comment]
Thanks for the update, @kgdiem ! Since your code is building up the string and parsing the data, that would need to be where you implement some kind of limits to prevent a DoS :)

**Comment by davidbnk** [comment]
Just FYI, the `Packets out of order` error might be related to `max_allowed_packet` mysql config

I was getting the `Packets out of order` error and when I increased `max_allowed_packet` the error stoped

**Comment by naterkane** [comment]
@davidbnk what did you increase `max_allowed_packet` from and to?

**Comment by davidbnk** [comment]
@naterkane increased to 16MB from the default of 1MB

**Comment by MrRobWad** [comment]
@davidbnk Thanks for your answer. I changed the default from 1MB to 16MB and started getting packet too large errors in place of the packets out of order errors. I proceeded to bump up a bit higher than your 16MB and was able to get the query to run.

**Comment by syrnick** [comment]
@kgdiem , your script is building a string and node has a hard limit of around 270MB for string length (it's hard-coded in v8 source). If you were hitting that limit, it would crash before calling `connection.query(insert);`

I don't see how this would trigger an error at mysql client unless the client does further string manipulation and then goes over the hard limit. It's typical to have rather cryptic error message when v8 string length limit is exceeded, because it's very unexpected.

side note, @kgdiem your example is vulnerable to SQL injection attack.

Does anyone has an isolated repro of this?

**Comment by kgdiem** [comment]
@syrnick I wasn't hitting the string limit because I was getting a packets out of order exception. 

@syrnick I was just reading data from a csv and dumping it into a DB to run queries on it :p.

**Comment by syrnick** [comment]
"I wasn't hitting the string limit because I was getting a packets out of order exception." that's a strong statement. Are you absolutely certain that the out of order exception is NOT caused by string limit?

**Comment by kgdiem** [comment]
I guess the only way to tell is by looking at the source!

**Comment by jochemstoel** [comment]
Hey I know it is years later but I experienced the same problem and it is caused by the database not accepting connections from the machine that is running your script. So if anyone is reading this in 2020 here it is.

The error:
```
Error: Packets out of order. Got: 1 Expected: 2
```

The solution:
```sql
GRANT ALL PRIVILEGES ON *.* TO 'username'@'%' WITH GRANT OPTION;
FLUSH PRIVILEGES;
``` 
See: [Stackoverflow on how to edit user access](https://stackoverflow.com/a/18821224/4696752)

**Comment by pablo-johnson** [comment]
I was having the same issue, i end up creating a new user in the db with host '% ' as value. The problem was my server was running in another ip (not 127.0.0.1 as the root user demands).
